app/controllers/coronavirus/sub_sections_controller.rb

```diff
@@ -9,11 +9,22 @@ def new
 
     def create
       @sub_section = page.sub_sections.new(sub_section_params)
-      if @sub_section.save && draft_updater.send
-        redirect_to coronavirus_page_path(page.slug), notice: "Sub-section was successfully created."
-      else
-        @sub_section.errors.add :base, draft_updater.errors.to_sentence
+
+      unless @sub_section.valid?
+        render :new
+        return
+      end
```

Comment:
I'm not sure if we're testing this conditional in the tests?

From what I can tell there's a bunch of code (too much in my opinion) that test the draft_updater.send fails with: https://github.com/alphagov/collections-publisher/blob/8a4d67ddcbe7d6a053ddab189099456f7fa5ba4c/spec/controllers/coronavirus/sub_sections_controller_spec.rb#L56-L84 but nothing for the situation where the model is invalid.

Similar situation with the update method. where it seems that the validation code is perhaps indirectly tested with https://github.com/alphagov/collections-publisher/blob/8a4d67ddcbe7d6a053ddab189099456f7fa5ba4c/spec/controllers/coronavirus/sub_sections_controller_spec.rb#L161-L167 but I'm not sure and seems a bit cryptic.

Reply:
So our thinking for this was that it would be covered by the featured link tests. That block is tested for create as well as update here https://github.com/alphagov/collections-publisher/blob/master/spec/controllers/coronavirus/sub_sections_controller_spec.rb#L98-L105. I do agree that this is slightly opaque but I'm happy enough to leave it that way until the work to tidy up the validation is done as I think it's important to be explicit about it once we're more knowingly relying on validation in the model.

The testing for the `flash["alert"] stuff is done here https://github.com/alphagov/collections-publisher/blob/master/spec/controllers/coronavirus/sub_sections_controller_spec.rb#L77-L79 because that's how that message is displayed unless I'm misunderstanding something.

Reply:
> So our thinking for this was that it would be covered by the featured link tests. That block is tested for create as well as update here https://github.com/alphagov/collections-publisher/blob/master/spec/controllers/coronavirus/sub_sections_controller_spec.rb#L98-L105. I do agree that this is slightly opaque but I'm happy enough to leave it that way until the work to tidy up the validation is done as I think it's important to be explicit about it once we're more knowingly relying on validation in the model.

Oh gosh, I didn't see that one. That does seem a bit cryptic.

It didn't seem like we'd have a particular motivation to fix this as part of the validation work so I think it's probably better sorted here. I quickly put together a commit that makes the 3 cases for each clear, if you're happy with it feel free to cherry pick it onto this PR: https://github.com/alphagov/collections-publisher/commit/6701bb0bd02443e55c8b3c8fc79c43c8eb06054d

> The testing for the `flash["alert"] stuff is done here https://github.com/alphagov/collections-publisher/blob/master/spec/controllers/coronavirus/sub_sections_controller_spec.rb#L77-L79 because that's how that message is displayed unless I'm misunderstanding something.

I was wondering more how that played out visually since it's a different way to set the error to how we did it before. It looks like it displays almost the same but skips the "There is a problem" line and is placed differently.

![Screenshot 2021-02-24 at 19 10 48](https://user-images.githubusercontent.com/282717/109052895-0db45b00-76d4-11eb-9f26-017df2821e8f.png)

It's worth checking with others, but this doesn't seem too big a concern for me.

One thing I have noticed though is that the `flash["alert"]` hangs around until the next page because we don't redirect. We'll need to change them to `flash.now["alert"]`.

For example: (note that it should be impossible to have validation errors and draft updater errors at the same time)

![Screenshot 2021-02-24 at 19 13 04](https://user-images.githubusercontent.com/282717/109053211-61bf3f80-76d4-11eb-959c-c4aaef398350.png)

Reply:
awesome, thanks for this, that break up of tests is lovely, we've put your commit in and adjusted the `flash` stuff. As for the difference in placement, I'm not too worried about this as it will all be brought into line when the model validation stuff happens.